src/indexer.c

```diff
@@ -0,0 +1,184 @@
+        RedisModule_DictIteratorStop(iter);
+        return 1;
+    }
+    RedisModule_DictIteratorStop(iter);
+    return 0;
+}
+
+int _difference(RedisModuleDict *left, RedisModuleDict *right, char **lastKey, size_t *lastKeySize) {
+    RedisModuleDictIter *iter;
+    if (*lastKey == NULL) {
+        iter = RedisModule_DictIteratorStartC(right, "^", NULL, 0);
+    } else {
+        iter = RedisModule_DictIteratorStartC(right, "^", lastKey, *lastKeySize);
+    }
+
+    char *currentKey;
+    size_t currentKeyLen;
+    while((currentKey = RedisModule_DictNextC(iter, &currentKeyLen, NULL)) != NULL) {
+        int doesNotExist = 0;
+        RedisModule_DictGetC(left, currentKey, currentKeyLen, &doesNotExist);
+        if (doesNotExist == 1) {
+            continue;
+        }
+        *lastKey = currentKey;
+        *lastKeySize = currentKeyLen;
+        RedisModule_DictDelC(left, currentKey, currentKeyLen, NULL);
+        RedisModule_DictIteratorStop(iter);
+        return 1;
+    }
+    RedisModule_DictIteratorStop(iter);
+    return 0;
+}
+
+RedisModuleDict * QueryIndexPredicate(RedisModuleCtx *ctx, QueryPredicate *predicate, RedisModuleDict *prevResults) {
+    RedisModuleDict *localResult = RedisModule_CreateDict(ctx);
+    RedisModuleCallReply *reply;
+    size_t _s;
+
+
+    if (predicate->type == NCONTAINS) {
+        RedisModuleString *index_key = RedisModule_CreateStringPrintf(ctx, "__index_%s",
+                RedisModule_StringPtrLen(predicate->label.key, &_s));
+        reply = RedisModule_Call(ctx, "SMEMBERS", "s", index_key);
+    } else {
+        const char *key = RedisModule_StringPtrLen(predicate->label.key, &_s);
+        const char *value = RedisModule_StringPtrLen(predicate->label.value, &_s);
+        RedisModuleString *index_key_value = RedisModule_CreateStringPrintf(ctx, "__index_%s=%s", key, value);
+        reply = RedisModule_Call(ctx, "SMEMBERS", "s", index_key_value);
+    }
+    size_t items = RedisModule_CallReplyLength(reply);
+
+    for (size_t j=0; j < items; j++) {
+        RedisModuleCallReply *item = RedisModule_CallReplyArrayElement(reply, j);
+        RedisModuleString *key_name = RedisModule_CreateStringFromCallReply(item);
+        RedisModule_DictSet(localResult, key_name, (void *)1);
+
+    }
+
+    if (reply)
+        RedisModule_FreeCallReply(reply);
+
+    if (prevResults != NULL) {
+        char *lastKey = NULL;
+        size_t lastKeySize;
+        if (predicate->type == EQ) {
+            while (_intersect(prevResults, localResult, &lastKey, &lastKeySize) != 0) {}
+        } else  if (predicate->type == NCONTAINS) {
+            while (_difference(prevResults, localResult, &lastKey, &lastKeySize) != 0) {}
+        } else if (predicate->type == NEQ){
+            while (_difference(prevResults, localResult, &lastKey, &lastKeySize) != 0) {}
+        }
+        return prevResults;
+    } else if (predicate->type == EQ) {
+        return localResult;
+    } else {
+        return prevResults;
```

Comment:
if you try to find all keys that don't match a label , and prevResults is NULL, this will return NULL instead of all the keys